### PR TITLE
Adding Flytekit via grayskull

### PR DESCRIPTION
**Pull request by bmcconeghy**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by bmcconeghy** [comment]
@asford tagging you here so we can _officially_ tick off the box that you're OK with being listed as a maintainer 😉

**Comment by asford** [comment]
Nice. I'm in.

@conda-forge/help-python This is ready for review.

**Comment by HeavyTony2** [comment]
The problem comes from protobuf 
https://github.com/conda-forge/protobuf-feedstock/pull/132
https://github.com/conda-forge/protobuf-feedstock/issues/133

File: recipes/flytekit/meta.yaml
Comment (asford) [inline recipes/flytekit/meta.yaml:13]:
```suggestion
  skip: true  # [py2k or win]
```

**Review by asford — COMMENTED** [review]
Probably just skip windows builds until proto unblocks?

**Comment by asford** [comment]
> The problem comes from protobuf 
> 
> https://github.com/conda-forge/protobuf-feedstock/pull/132
> 
> https://github.com/conda-forge/protobuf-feedstock/issues/133

Thanks! Can you rereview the skip logic?

File: recipes/flytekit/meta.yaml
Comment (HeavyTony2) [reply recipes/flytekit/meta.yaml:13]: You could add notes under the line like
#Skipping Windows for now
#xref: https://github.com/conda-forge/protobuf-feedstock/issues/133

**Comment by asford** [comment]
@HeavyTony2 @conda-forge/help-python This is ready for re-review.

**Comment by bmcconeghy** [comment]
@conda-forge/help-python This should be ready for review. Skipped the windows build for now.

**Review by asford — APPROVED** [review]

**Review by HeavyTony2 — APPROVED** [review]

**Comment by asford** [comment]
@conda-forge/help-python I think this needs a final staged-recipes review for the merge?

File: recipes/flyteidl/meta.yaml
Comment (BastianZim) [inline recipes/flyteidl/meta.yaml:2]: The newest version is `0.21.25`. You can go with this one and the bot should update it but just fyi.

File: recipes/flyteidl/meta.yaml
Comment (BastianZim) [inline recipes/flyteidl/meta.yaml:38]:
```suggestion
  license_file:
    - LICENSE
    - NOTICE
```

File: recipes/flyteidl/LICENSE
Comment (BastianZim) [inline recipes/flyteidl/LICENSE:1]: This is already included in the tar ball.

File: recipes/checksumdir/LICENSE
Comment (BastianZim) [inline recipes/checksumdir/LICENSE:1]: Where does this license come from? Upstream lists this one, which is included in the tar ball: https://github.com/cakepietoast/checksumdir/blob/master/LICENSE.txt

File: recipes/flytekit/LICENSE
Comment (BastianZim) [inline recipes/flytekit/LICENSE:1]: This license is also included in the tar ball.

File: recipes/flytekit/meta.yaml
Comment (BastianZim) [inline recipes/flytekit/meta.yaml:15]: Is the only reason for this not being `noarch` the skip? In that case you can leave it `noarch` and restrict it by using the `__unix` virtual package.

File: recipes/singledispatchmethod/LICENSE
Comment (BastianZim) [inline recipes/singledispatchmethod/LICENSE:1]: Already included in tar ball.

File: recipes/singledispatchmethod/meta.yaml
Comment (BastianZim) [inline recipes/singledispatchmethod/meta.yaml:23]: Conda-forge only supports `Python >=3.7` for non `noarch` packages so you can remove that and make the minimum python version 3.4 and make this package `noarch`.

**Review by BastianZim — CHANGES_REQUESTED** [review]

File: recipes/flytekit/meta.yaml
Comment (BastianZim) [reply recipes/flytekit/meta.yaml:15]: Edit: Scratch that, just saw that you have `singledispatchmethod  # [py<38]` so it cannot be `noarch`.

File: recipes/flyteidl/meta.yaml
Comment (bmcconeghy) [reply recipes/flyteidl/meta.yaml:2]: Thanks for the catch. Must have _just_ updated after I put this together on the 28th.

File: recipes/checksumdir/LICENSE
Comment (bmcconeghy) [reply recipes/checksumdir/LICENSE:1]: Yeah, sorry about that, I'll remove it and reference the `LICENSE.txt` within the tarball instead

File: recipes/flyteidl/LICENSE
Comment (bmcconeghy) [reply recipes/flyteidl/LICENSE:1]: Gotcha. Removed the file.

File: recipes/flytekit/LICENSE
Comment (bmcconeghy) [reply recipes/flytekit/LICENSE:1]: Gotcha. Removed the file.

File: recipes/singledispatchmethod/LICENSE
Comment (bmcconeghy) [reply recipes/singledispatchmethod/LICENSE:1]: Gotcha. Removed the file.

File: recipes/checksumdir/LICENSE
Comment (BastianZim) [reply recipes/checksumdir/LICENSE:1]: No worries. Generally, Licenses from the tar ball should be used and if they’re not included you can add them manually and make a PR upstream to include them with the next release.

File: recipes/singledispatchmethod/meta.yaml
Comment (bmcconeghy) [reply recipes/singledispatchmethod/meta.yaml:23]: OK sounds good. By "remove that", do you mean remove the line `- singledispatch  # [py<34]` or just the `# [py<34]`? Sorry, newb contributor here!

Comment (BastianZim) [reply recipes/singledispatchmethod/meta.yaml:23]: Yes sorry should've been more clear. 
```suggestion
```

Basically, when a package has a version selector on a dependency it needs to be built for every python version separately. Since conda-forge only builds for python3.7, 3.8, 3.9 and 3.10, you'll never hit python3.3 and need this package.

When a package is noarch, it can be built once and installed under every python version that is allowed in your dependencies. 
Since Python3.3 requires this package, your minimum version should therefore be anything greater than that.

_Note_: I haven't checked if upstream has declared a higher python boundary than that anywhere else yet.

Comment (bmcconeghy) [reply recipes/singledispatchmethod/meta.yaml]: Great, thanks for the clarification.

**Comment by bmcconeghy** [comment]
@BastianZim OK, I think we're good now?

File: recipes/flyteidl/meta.yaml
Comment (BastianZim) [inline recipes/flyteidl/meta.yaml:23]: Sorry, missed that the first time. Similarly to https://github.com/conda-forge/staged-recipes/pull/17778#discussion_r797990317 you can remove this as conda-forge cannot build `python2.7`, make the package `noarch` and restrict the lowest python version to whatever upstream lists as the lowest version that is larger than `2.7`.

File: recipes/flytekit/meta.yaml
Comment (BastianZim) [inline recipes/flytekit/meta.yaml:81]:
```suggestion
  license_file: 
    - LICENSE
    - NOTICE
```

The NOTICE file is [listed](https://github.com/flyteorg/flytekit/blob/master/NOTICE) in the repo but not distributed with the tar ball. I would make a PR upstream that includes it in the tar ball and include the file manually here until that is merged and released.

**Review by BastianZim — CHANGES_REQUESTED** [review]
Great progress! I just added a couple of small things that I missed in the first run but that should be all.

File: recipes/flytekit/meta.yaml
Comment (bmcconeghy) [reply recipes/flytekit/meta.yaml:81]: OK, will do.

File: recipes/singledispatchmethod/meta.yaml
Comment (bmcconeghy) [reply recipes/singledispatchmethod/meta.yaml]: Yes, I think you are correct!

File: recipes/flyteidl/meta.yaml
Comment (bmcconeghy) [reply recipes/flyteidl/meta.yaml:23]: OK, will do. Couple questions though:

1. By upstream, do you mean `flytekit`? This is the main package of interest for us. I don't think it directly requires a specific version of python, unless I'm misinterpreting.
2. Do we even need to bring in `singledispatchmethod`, if we set the python version to `>=3.7`? This would then nullify the need to specify this `typing` version constraint, right?

Comment (BastianZim) [reply recipes/flyteidl/meta.yaml:23]: RE 1: I meant `flyteidl`. Basically, you'll need to figure out if they have a minimum python version they support and then add that to the `host` and `run` python, similarly to what you did with `singledispatchmethod`. If they have none, `python>=3.6` is a safe bet.

RE 2: If you set the minimum version of `flytekit` to `python>=3.8` (Not `python>=3.7` since it applies to everything lower than `python3.8`) then no, you wouldn't need `singledispatchmethod`. That is up to you as maintainer of the conda-forge distribution. That would also mean that you can probably switch to `noarch` with that package although I haven't had a look at your windows issue so that is no definitive.

Comment (BastianZim) [reply recipes/flyteidl/meta.yaml:23]: On a side-note: On conda-forge `typing` is an empty package as we don't support `python2.7` so it wouldn't do anything anyways.

Ref: https://conda-forge.org/docs/maintainer/knowledge_base.html#empty-python-packages

Comment (asford) [reply recipes/flyteidl/meta.yaml]: 👍 I'd be in favor of just bumping to py 3.8+ and going noarch for the time being. That'll make post-merge dance of bring up builds easier.

Comment (bmcconeghy) [reply recipes/flyteidl/meta.yaml]: OK, I'll change `flytekit` to a noarch build and bump the version to 3.8+. Shall I also remove the port in of `singledispatchmethod`?

Comment (asford) [reply recipes/flyteidl/meta.yaml]: For sure, I'm ready to burn the boats on py38 anyway. 😉

Comment (bmcconeghy) [reply recipes/flyteidl/meta.yaml]: OK sweet, same here!

**Comment by BastianZim** [comment]
If the only reason you’re reverting `noarch` is a missing dependency on windows then that’s not a problem; the CI can fail because of that. 

See the note here: https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python

Also, you don’t need to skip appveyor, we’re not using that anyone. We just forgot to remove it from the docs. Do you know where you saw that so that we can remove it?

**Comment by bmcconeghy** [comment]
> If the only reason you’re reverting `noarch` is a missing dependency on windows then that’s not a problem; the CI can fail because of that.
> 
> See the note here: https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python

I'm not sure if it's a missing dependency, or some other issue, but it is mentioned [here](https://github.com/conda-forge/protobuf-feedstock/issues/133) and is an issue with protobuf on windows.

> Also, you don’t need to skip appveyor, we’re not using that anyone. We just forgot to remove it from the docs. Do you know where you saw that so that we can remove it?

Yes, I read this tip on the main landing page, so it's referenced in the README [here](https://github.com/conda-forge/staged-recipes/blob/main/README.md#3-how-do-i-exclude-a-platform).

File: recipes/flytekit/meta.yaml
Comment (BastianZim) [inline recipes/flytekit/meta.yaml:15]:
```suggestion
  noarch: python
```

File: recipes/flytekit/meta.yaml
Comment (BastianZim) [inline recipes/flytekit/meta.yaml:28]:
```suggestion
    - python >=3.8
```

File: recipes/flytekit/meta.yaml
Comment (BastianZim) [inline recipes/flytekit/meta.yaml:48]:
```suggestion
    - python >=3.8
```

File: recipes/flytekit/meta.yaml
Comment (BastianZim) [inline recipes/flytekit/meta.yaml:61]:
```suggestion
    - wrapt >=1.0.0,<2.0.0
    - __unix
    # xref: conda-forge/protobuf-feedstock#133
```

Ref: https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html

**Review by BastianZim — CHANGES_REQUESTED** [review]
> I'm not sure if it's a missing dependency, or some other issue, but it is mentioned https://github.com/conda-forge/protobuf-feedstock/issues/133 and is an issue with protobuf on windows.

Ok in that case it can still be `noarch` and you can just restrict the platform the code can be used on by using a virtual package: https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html
This makes it much easier for you since you don't need to build for every platform.

> Yes, I read this tip on the main landing page, so it's referenced in the README [here](https://github.com/conda-forge/staged-recipes/blob/main/README.md#3-how-do-i-exclude-a-platform).

Thanks!

**Comment by bmcconeghy** [comment]
> > I'm not sure if it's a missing dependency, or some other issue, but it is mentioned https://github.com/conda-forge/protobuf-feedstock/issues/133 and is an issue with protobuf on windows.
> 
> Ok in that case it can still be `noarch` and you can just restrict the platform the code can be used on by using a virtual package: https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html
> This makes it much easier for you since you don't need to build for every platform.
> 
> > Yes, I read this tip on the main landing page, so it's referenced in the README [here](https://github.com/conda-forge/staged-recipes/blob/main/README.md#3-how-do-i-exclude-a-platform).
> 
> Thanks!

Great! Since we don't care if it is compatible with Windows, would I use `__win` or `__unix`? For some reason, I'm not grasping the concept of virtual packages. Apologies!

**Comment by BastianZim** [comment]
No worries. You'd need `__unix` one.
Basically, the virtual package is available on the platform it says (`__linux` is available when conda is running on Linux, `__osx` is available when conda is running on macOS etc.). So when someone installs the package, the conda solver will check if the `__unix`(Linux and macOS) package can be found which is not the case on Windows and informs the user that the package cannot be installed on this architecture.

**Comment by BastianZim** [comment]
Just run `conda info` on your machine and it should list all virtual packages. If a package then lists a virtual package under run requirements that is not in this list, you cannot install it.

**Comment by bmcconeghy** [comment]
Hmm... Seems like the `win_64` [CI is failing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=453506&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b&l=3604)

**Comment by BastianZim** [comment]
No this is expected and fine. Since we're restricting this to unix, win will never succeed.

**Comment by BastianZim** [comment]
Let me just ping the team for a final review.

@conda-forge/staged-recipes Could someone please take a final look, thanks!

**Comment by bmcconeghy** [comment]
Thank you @xhochy !

**Pull request merged**
